wurst/systems/spells/mage/dementia/DarkGateNew.wurst

```diff
@@ -0,0 +1,405 @@
+    angle orbAngle
+    unit ownerUnit
+    bool isChanneling
+    real portalRadius
+    real elapsedLifetime
+    real channeledTime
+    real lastSpellCastTimestamp
+    real spellIntervalCurrent
+    real durationBuildup
+    vec3 centerPos
+    CallbackCounted portalActiveLoop = null
+    CallbackCounted portalCreationLoop = null
+
+    LinkedList<PortalOrb> orbs = new LinkedList<PortalOrb>()
+    LinkedList<Summon> summons = new LinkedList<Summon>()
+
+    construct (unit caster, vec3 pos)
+        ownerUnit = caster
+        isChanneling = true
+        portalRadius = 50
+        centerPos = pos
+        elapsedLifetime = 0
+        channeledTime = 0
+        durationBuildup = 0
+        lastSpellCastTimestamp = -9999
+        orbAngle = angle(0)
+        spellIntervalCurrent = SPELLCAST_INTERVAL_START
+        startPortalCreation()
+
+    function startPortalCreation()
+        for i = 1 to ORB_COUNT
+            orbs.push(new PortalOrb(this, centerPos))
+        
+        portalCreationLoop = doPeriodicallyTimed(ORB_LOOP_INTERVAL, GROW_DURATION) portalGrowLoop ->
+            if not isChanneling or portalGrowLoop.isLast()
+                portalGrowLoop.stop()
+                this.startPortalLoop()
+            else
+                portalGrowLoopUpdate()
+
+    function portalGrowLoopUpdate()
+        //Avoid creating vision on every loop cycle with mod check
+        if (portalCreationLoop.getCount() mod 30 == 0)
+            var vis = createVision(ownerUnit.getOwner(), centerPos.toVec2(), portalRadius+100, true)
+            vis.start()
+            //Should be the full duration of portal calculated below for vision mods destruction
+            doAfter(durationBuildup + GROW_DURATION - elapsedLifetime) ->
+                vis.destr()
+        
+        //Update things that change with channeling time
+        spellIntervalCurrent = SPELLCAST_INTERVAL_START - (SPELLCAST_INTERVAL_START - SPELLCAST_INTERVAL_FASTEST) * (elapsedLifetime/GROW_DURATION)
+        elapsedLifetime += ORB_LOOP_INTERVAL
+        channeledTime += ORB_LOOP_INTERVAL
+        portalRadius += ORB_LOOP_INTERVAL * PORTAL_GROWSPEED
+        durationBuildup = (channeledTime / GROW_DURATION) * DURATION_MAX
+        orbAngle = angle(orbAngle.radians() + ORB_LOOP_INTERVAL * ORB_ANGULARSPEED_SUMMON * DEGTORAD)
+
+        //Move orbs around
+        updateOrbAngles()
+
+    function isUnitInsidePortal(unit u) returns boolean
+        return this.centerPos.toVec2().distanceTo(u.getPos()) <= this.portalRadius
+
+    function randomOrbCastSpell()
+        var orb = this.orbs.get( GetRandomInt(0, orbs.size() - 1))
+        var centerPosVec2 = centerPos.toVec2()
+        var _foundTarget = false
+
+        //Find random target and cast + spawn effect on orb
+        forUnitsInRange(centerPosVec2, portalRadius) t ->
+            if not _foundTarget
+                if isValidEnemyTarget(t)
+                    orb.castRandomSpell(t)
+                    _foundTarget = true
```

Comment:
I wonder if we should use something that may be *more* random like what I have below. `forUnitsInRange` may be deterministic and thus we'll always cast on the same unit.

```
let targets = CreateGroup..enumUnitsInRect(pos, range, Filter(-> isValidEnemyTarget(GetFilterUnit()))
let target = targets.getRandomUnit()
targets.destr()
```

Reply:
Good point, in testing the current code SEEMED random so this never came up but that would ensure randomness.